src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs

```diff
@@ -247,6 +250,111 @@ public static RecordValue Execute()
+        [InlineData("{x: {x : 2}}", false)]
+        public void CustomFunctionRecordTest(string inputRecord, bool isSuccess)
+        {
+            var recordType = RecordType.Empty().Add(new NamedFormulaType("x", FormulaType.Number)).Add(new NamedFormulaType("y", FormulaType.Number));
+            var tableType = recordType.ToTable();
+            
+            CustomFunctionRecord(new TestRecordsCustomFunction(recordType), "(" + inputRecord + ")", isSuccess);
+            CustomFunctionRecord(new TestTableCustomFunction(tableType), "([" + inputRecord + "])", isSuccess);
+        }
+
+        private void CustomFunctionRecord(ReflectionFunction reflectionFunction, string inputRecord, bool isSuccess)
+        {
+            var config = new PowerFxConfig();
+            config.AddFunction(reflectionFunction);
+            
+            var engine = new RecalcEngine(config);
+            var func = engine.GetFunctionsByName(reflectionFunction.GetFunctionName());
+            Assert.NotNull(func);
+
+            var check = engine.Check(reflectionFunction.GetFunctionName() + inputRecord);
+
+            FormulaValue result = null; 
+            ConfiguredTaskAwaitable<FormulaValue> resultAsync;
+            string errorMsg = string.Empty;
+            var rc = new RuntimeConfig();
+
+            try
+            {
+                result = check.GetEvaluator().Eval(rc);
+                resultAsync = check.GetEvaluator().EvalAsync(CancellationToken.None, rc).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                errorMsg = ex.ToString();
+            }
+
+            if (isSuccess)
+            {
+                Assert.True(check.IsSuccess);                
+                Assert.IsNotType<ErrorValue>(result);
+                Assert.IsNotType<ErrorValue>(resultAsync);
+            }
+            else
+            {
+                Assert.False(check.IsSuccess);
+                Assert.Contains("invalid arguments", errorMsg);
+            }
+        }
+
+        // Must have "Function" suffix. 
+        private class TestRecordsCustomFunction : ReflectionFunction
+        {
+            public TestRecordsCustomFunction(RecordType recordType)
+                : base("RecordsTest", recordType, recordType)
+            {
+            }
+
+            // Must have "Execute" method. 
+            public static RecordValue Execute(RecordValue recordValue)
+            {
+                return recordValue;
+            }
+        }
+
+        // Must have "Function" suffix. 
+        private class TestTableCustomFunction : ReflectionFunction
+        {
+            public TestTableCustomFunction(TableType tableType)
+                : base("RecordsTest", tableType, tableType)
+            {
+            }
+
+            // Must have "Execute" method. 
+            public static TableValue Execute(TableValue tableValue)
+            {
+                return tableValue;
+            }
+        }
```

Comment:
You can merge those two functions into a single one:

        // Must have "Function" suffix. 
        private class TestAggregateIdentityCustomFunction<TType, TValue> : ReflectionFunction
            where TType : AggregateType
            where TValue : FormulaValue
        {
            public TestAggregateIdentityCustomFunction(TType argType)
                : base("RecordsTest", argType, argType)
            {
            }

            // Must have "Execute" method. 
            public static TValue Execute(TValue recordValue)
            {
                return recordValue;
            }
        }

Reply:
updated.